### PR TITLE
GA sampling overrides

### LINKED ISSUE #3433: General Availability target for sampling overrides
The sampling override feature for Application Insights for Java is in preview as of December 2023.
https://learn.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-sampling-overrides

Please keep us informed of the target for general availability and any changes we may need to make to switch to the GA version if currently using the preview version. Thanks!

### PR DESCRIPTION
**Pull request by heyams**
Fix #3433

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/ConfigurationBuilder.java [hunk 1]
```diff
@@ -191,7 +191,13 @@ private static void logConfigurationWarnings(Configuration config) {
...
     }
-    for (SamplingOverride override : config.preview.sampling.overrides) {
+
+    if (!config.preview.sampling.overrides.isEmpty()) {
+      configurationLogger.warn(
+          "\"Sampling overrides\" is no longer in preview and it has been GA since 3.5 GA,");
+      config.sampling.overrides = config.preview.sampling.overrides;
+    }
+    for (SamplingOverride override : config.sampling.overrides) {
       if (override.telemetryKind != null) {
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/ConfigurationBuilder.java:201]: @trask probably can drop supported deprecated 'sampling overrides' attributes at this point?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java [hunk 2]
```diff
@@ -192,7 +194,7 @@ public static class SamplingPreview {
     // future goal: make parentBased sampling the default if item count is received via tracestate
     public boolean parentBased;
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java:195]: what about this? should we GA this in the same PR?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 3]
```diff
@@ -349,6 +349,10 @@ private static Set<Feature> initStatsbeatFeatureSet(Configuration config) {
     if (!config.preview.sampling.overrides.isEmpty()) {
       featureList.add(Feature.PREVIEW_SAMPLING);
     }
+    // TODO (heya) track sampling overrides GA feature after updating the standalone exporter
+    //    if (!config.sampling.overrides.isEmpty()) {
+    //      featureList.add(Feature.SAMPLING);
+    //    }
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:355]: can go ahead and send PR to azure-sdk-for-java repo to rename `PREVIEW_SAMPLING` -> `SAMPLING_OVERRIDES`

Comment (trask) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:355]:
```suggestion
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 4]
```diff
@@ -349,6 +349,10 @@ private static Set<Feature> initStatsbeatFeatureSet(Configuration config) {
     if (!config.preview.sampling.overrides.isEmpty()) {
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:349]:
```suggestion
    if (!config.sampling.overrides.isEmpty()) {
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java [hunk 5]
```diff
@@ -28,11 +28,11 @@ public static Sampler getSampler(
     }
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/Samplers.java:28]: can we remove the `Configuration.SamplingPreview`?

can you check for any other references to `SamplingPreview`?

File: smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/ExceptionSamplingOverridesTest.java [hunk 6]
```diff
@@ -15,23 +15,36 @@
...
-    assertThat(testing.mockedIngestion.getCountForType("ExceptionData")).isZero();
+    // Exception from the log will get sampled out and exception from the request won't.
+    List<ExceptionData> exceptions =
+        testing.mockedIngestion.getTelemetryDataByTypeInRequest("ExceptionData");
+    assertThat(exceptions.size()).isEqualTo(1);
+    assertThat(exceptions.get(0).getProperties().size()).isEqualTo(0);
+    ExceptionDetails exceptionDetails = exceptions.get(0).getExceptions().get(0);
+    assertThat(exceptionDetails.getStack()).isNotNull();
+    assertThat(exceptionDetails.getTypeName()).isEqualTo("java.lang.RuntimeException");
+    assertThat(exceptionDetails.getMessage()).isEqualTo("this is an expected exception");
```
Comment (heyams) [inline smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/ExceptionSamplingOverridesTest.java:47]: Exception will be zero when https://github.com/Azure/azure-sdk-for-java/pull/38687 is released.

**Review by trask — COMMENTED** [review]
can you move the exception related smoke tests over to #3555?

File: smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/ExceptionWithSamplingOverridesTest.java [hunk 7]
```diff
@@ -19,10 +19,12 @@
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
-@UseAgent("applicationinsights3.json")
-abstract class ExceptionSamplingOverridesTest {
+@UseAgent("applicationinsights-exception-with-sampling-overrides.json")
+abstract class ExceptionWithSamplingOverridesTest {
```
Comment (trask) [inline smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/ExceptionWithSamplingOverridesTest.java:23]: can you revert these changes since they are in the other PR now?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java [hunk 8]
```diff
@@ -156,6 +156,7 @@ public static class Role {
   }
 
   public static class Sampling {
+    public boolean parentBased;
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/configuration/Configuration.java:159]: I don't think we discussed GA'ing "parent based"?

File: smoke-tests/apps/SpringBoot/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootTest.java [hunk 9]
```diff
@@ -60,26 +58,11 @@ void testResultCodeWhenRestControllerThrows() throws Exception {
...
-              }
-              // lastly, filter out ExceptionData captured from tomcat logger
-              ExceptionData data = (ExceptionData) ((Data<?>) input.getData()).getBaseData();
-              return !data.getProperties().containsKey("LoggerName");
-            },
-            1,
-            10,
-            TimeUnit.SECONDS);
+    List<Envelope> exceptions = testing.mockedIngestion.waitForItems("ExceptionData", 1);
+    assertThat(exceptions).hasSize(1);
```
Comment (trask) [inline smoke-tests/apps/SpringBoot/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SpringBootTest.java:62]: aren't these changes in the other PR?

File: smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/ExceptionSamplingOverridesTest.java [hunk 10]
```diff
@@ -22,7 +22,9 @@
 @UseAgent("applicationinsights3.json")
 abstract class ExceptionSamplingOverridesTest {
 
-  @RegisterExtension static final SmokeTestExtension testing = SmokeTestExtension.create();
+  @RegisterExtension
+  static final SmokeTestExtension testing =
+      SmokeTestExtension.builder().setSelfDiagnosticsLevel("DEBUG").build();
```
Comment (trask) [inline smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/ExceptionSamplingOverridesTest.java:27]: can revert

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 11]
```diff
@@ -400,9 +400,10 @@ private static Set<Feature> initStatsbeatFeatureSet(Configuration config) {
     if (config.preview.browserSdkLoader.enabled) {
       featureList.add(Feature.BROWSER_SDK_LOADER);
     }
-    if (!config.preview.sampling.overrides.isEmpty()) {
-      featureList.add(Feature.PREVIEW_SAMPLING);
-    }
+    // TODO (heya) track sampling overrides GA feature after updating the standalone exporter
+    //    if (!config.sampling.overrides.isEmpty()) {
+    //      featureList.add(Feature.SAMPLING);
+    //    }
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:406]: can fix this after merging #3555 and rebasing this PR

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/diagnostics/etw/EtwAppender.java [hunk 12]
```diff
@@ -45,7 +45,7 @@ public void start() {
       this.etwProvider.writeEvent(event);
     } catch (LinkageError | ApplicationInsightsEtwException e) {
       String message = "EtwProvider failed to initialize.";
-      LoggerFactory.getLogger(DiagnosticsHelper.DIAGNOSTICS_LOGGER_NAME).error(message, e);
+      LoggerFactory.getLogger(DiagnosticsHelper.DIAGNOSTICS_LOGGER_NAME).debug(message, e);
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/diagnostics/etw/EtwAppender.java:48]: 👍

**Review by trask — APPROVED** [review]
